components/tidb_query/src/codec/mysql/time/mod.rs

```diff
@@ -1358,6 +1364,50 @@ impl Time {
     pub fn invalid_zero(self) -> bool {
         self.month() == 0 || self.day() == 0
     }
+
+    pub fn from_days(ctx: &mut EvalContext, daynr: u32) -> Result<Self> {
+        let (year, month, day) = Time::get_date_from_daynr(daynr);
+        let time_args = TimeArgs {
+            year,
+            month,
+            day,
+            ..Default::default()
+        };
+        Time::new(ctx, time_args)
+    }
+
+    // Changes a daynr to year, month and day, daynr 0 is returned as date 00.00.00
+    #[inline]
+    fn get_date_from_daynr(daynr: u32) -> (u32, u32, u32) {
```

Comment:
Interesting implementation. How is the performance compared to TiDB's implementation?

Reply:
Two implementations and benchmark tests:
```Rust
impl Time {
    // Changes a daynr to year, month and day, daynr 0 is returned as date 00.00.00
    #[inline]
    pub fn get_date_from_daynr(daynr: u32) -> (u32, u32, u32) {
        if daynr <= 365 || daynr >= 3_652_425 {
            return (0, 0, 0);
        }

        let (mut year, mut month, mut days) = (1, 1, daynr);
        days -= 365;
        for each_year in 1.. {
            let days_in_year = if is_leap_year(each_year) { 366 } else { 365 };
            if days > days_in_year {
                year += 1;
                days -= days_in_year;
            } else {
                break;
            }
        }
        for each_month in 1..=12 {
            let last_day_of_month = last_day_of_month(year, each_month);
            if days > last_day_of_month {
                month += 1;
                days -= last_day_of_month;
            } else {
                break;
            }
        }

        let day = days;

        (year, month, day)
    }

    #[inline]
    pub fn get_date_from_daynr2(daynr: u32) -> (u32, u32, u32) {
        if daynr <= 365 || daynr >= 3_652_425 {
            return (0, 0, 0);
        }

        let mut year = daynr * 100 / 36525;
        let temp = (((year - 1) / 100 + 1) * 3) / 4;
        let mut day_of_year = daynr - year * 365 - (year - 1) / 4 + temp;

        let mut days_in_year = if is_leap_year(year) { 366 } else { 365 };
        while day_of_year > days_in_year {
            day_of_year -= days_in_year;
            year += 1;
            days_in_year = if is_leap_year(year) { 366 } else { 365 };
        }

        let mut leap_day = 0;
        if days_in_year == 366 && day_of_year > 31 + 28 {
            day_of_year -= 1;
            if day_of_year == 31 + 28 {
                leap_day = 1;
            }
        }

        let mut month = 1;
        let days_in_month = vec![31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        for days in days_in_month {
            if day_of_year <= days {
                break;
            }
            day_of_year -= days;
            month += 1;
        }

        let days = day_of_year + leap_day;

        (year, month, days)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[bench]
    fn bench_from_days(b: &mut Bencher) {
        b.iter(|| {
            let cases = vec![
                (140, (0, 0, 0)),          // mysql FROM_DAYS returns 0000-00-00 for any day <= 365.
                (735_000, (2012, 05, 12)), // Leap year.
                (735_030, (2012, 06, 11)),
                (735_130, (2012, 09, 19)),
                (734_909, (2012, 02, 11)),
                (734_878, (2012, 01, 11)),
                (734_927, (2012, 02, 29)),
                (734_634, (2011, 05, 12)), // Non Leap year.
                (734_664, (2011, 06, 11)),
                (734_764, (2011, 09, 19)),
                (734_544, (2011, 02, 11)),
                (734_513, (2011, 01, 11)),
                (3_652_424, (9999, 12, 31)),
                (3_652_425, (0000, 00, 00)), // mysql FROM_DAYS returns 0000-00-00 for any day >= 3652425
            ];
            for (arg, exp) in cases {
                assert_eq!(Time::get_date_from_daynr(arg), exp);
            }
        });
    }

    #[bench]
    fn bench_from_days2(b: &mut Bencher) {
        b.iter(|| {
            let cases = vec![
                (140, (0, 0, 0)),          // mysql FROM_DAYS returns 0000-00-00 for any day <= 365.
                (735_000, (2012, 05, 12)), // Leap year.
                (735_030, (2012, 06, 11)),
                (735_130, (2012, 09, 19)),
                (734_909, (2012, 02, 11)),
                (734_878, (2012, 01, 11)),
                (734_927, (2012, 02, 29)),
                (734_634, (2011, 05, 12)), // Non Leap year.
                (734_664, (2011, 06, 11)),
                (734_764, (2011, 09, 19)),
                (734_544, (2011, 02, 11)),
                (734_513, (2011, 01, 11)),
                (3_652_424, (9999, 12, 31)),
                (3_652_425, (0000, 00, 00)), // mysql FROM_DAYS returns 0000-00-00 for any day >= 3652425
            ];
            for (arg, exp) in cases {
                assert_eq!(Time::get_date_from_daynr2(arg), exp);
            }
        });
    }
}
```

Benchmark outputs:
```
# cargo bench
test codec::mysql::time::tests::bench_from_days                                           ... bench:      38,366 ns/iter (+/- 2,759)
test codec::mysql::time::tests::bench_from_days2                                          ... bench:         108 ns/iter (+/- 7)
```
Actually the implementation 1 is much slower than  2 (~300x) :(

Reply:
My implementation uses iteration to compute ‘year’ and ‘day_of_year’, while the TiDB’s(and MySQL’s) uses math. The latter is much faster.